api/src/main/java/org/openmrs/module/reporting/report/processor/HttpReportProcessor.java

```diff
@@ -0,0 +1,64 @@
+package org.openmrs.module.reporting.report.processor;
+
+import java.io.OutputStreamWriter;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.module.reporting.report.Report;
+import org.springframework.stereotype.Component;
+
+@Component
+public class HttpReportProcessor implements ReportProcessor {
+
+	protected Log log = LogFactory.getLog(this.getClass());
+
+	/**
+	 * @see ReportProcessor#getConfigurationPropertyNames()
+	 */
+	@Override
+	public List<String> getConfigurationPropertyNames() {
+		List<String> ret = new ArrayList<String>();
+		ret.add("connection url");
+		ret.add("subject");
+		ret.add("Add report");
+		return ret;
+	}
+
+	/**
+	 * Performs some action on the given report
+	 * 
+	 * @param report the Report to process
+	 */
+	@Override
+	public void process(Report report, Properties configuration) {
+		try {
+			URL url = new URL(configuration.getProperty("connection url"));
+			HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+			connection.setRequestMethod("POST");
+			connection.setRequestProperty("Content-Type", "text/html; charset=UTF-8");
+			connection.setDoOutput(true);
+			connection.connect();
+			 
+			if (report.getRenderedOutput() != null && "true".equalsIgnoreCase(configuration.getProperty("Add report"))) {	
+				String addcontent = configuration.getProperty(("Add report") + configuration.getProperty("subject"));
+				if (report.getOutputContentType().contains("text") || report.getOutputContentType().contains("html")) {
+					addcontent =  new String(report.getRenderedOutput(),"UTF-8");	
+				}	
+				OutputStreamWriter writer = new OutputStreamWriter(
+					    connection.getOutputStream());
+				writer.write(addcontent);
+				writer.flush();
+				writer.close();
```

Comment:
Does this need to be in a finally block?

Reply:
No because the finally blocks can't be inserted in try-catch blocks

Reply:
With I/O, typically it's a good idea to close the connection in a finally block.  You can have nested try/catch blocks if you need to.  In this case, I would declare "HttpURLConnection" outside of the try block, assign it and use it inside the try block, and close it in the finally block:

```
HttpURLConnection connection;
OutputStreamWriter writer;
try {
  connection = (HttpURLConnection) url.openConnection();
  ...
  writer = new OutputStreamWriter(connection.getOutputStream());
  ...
}
catch (Exception e) {
  throw new RuntimeException("Error occurred while sending report via HTTP POST", e);
}
finally {
  IOUtils.closeQuietly(writer);
  IOUtils.closeQuietly(connection);
}
```

Reply:
@mseaton the connection variable is as a result of casting the `url` variable to `HttpURLConnection` of which if am to declare `"HttpURLConnection"` outside of the try block then i will need to declare `URL ` outside to, which requires also a try block to catch the `MalformedURLException` and if this is done,then the scope of the `url` variable used here `url.openConnection()` will be limited